### PR TITLE
Add an on-device RNG engine class and a couple of distributions

### PR DESCRIPTION
**Pull request by amandalund**

### CODE REVIEW

File: src/random/GenerateCanonical.hh [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    explicit CELER_INLINE_FUNCTION GenerateCanonical() { }
+
+    // Sample a random number
+    CELER_INLINE_FUNCTION result_type operator()(Generator& rng);
+};
+
+} // namespace celeritas
+
+#ifdef __NVCC__
+#    include "random/RngEngine.cuh"
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/random/GenerateCanonical.hh:42]: Agreed, will do.

File: src/random/UniformRealDistribution.i.hh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * \brief Sample a random number according to the distribution
+ */
+template<class T>
+template<class Generator>
+CELER_INLINE_FUNCTION auto
+UniformRealDistribution<T>::operator()(Generator& rng) -> result_type
+{
+    //GenerateCanonical<Generator, T> sample_uniform;
+    //return delta_ * sample_uniform(rng) + a_;
+    return delta_ * GenerateCanonical<Generator, T>()(rng) + a_;
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/random/UniformRealDistribution.i.hh:36]: Oops, didn't realize I'd left that in there -- sounds good, I'll switch it.

File: src/random/RngStateView.cuh [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    //! Number of states
+    __device__ size_type size() const { return data_.size; }
+
+    //! Get a reference to the local state for a thread
+    __device__ RngEngine operator[](size_type idx) const
+    {
+        REQUIRE(idx < this->size());
+        return RngEngine(data_.rng + idx);
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/random/RngStateView.cuh]: I modified `RngStateView` and `RngStateContainer` based on what we discussed offline, so the view is now a simple struct and the container is a `.hh` file.

File: src/random/RngEngine.cuh [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  public:
+    // Construct from state
+    __device__ inline RngEngine(RngState* state);
+
+    // Sample a random number
+    __device__ inline result_type operator()();
+
+    // Sample a random number uniformly distributed on (0.0, 1.0]
+    template<class ResultType = double>
+    __device__ inline ResultType sample_uniform();
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/random/RngEngine.cuh]: Done.

**Comment by sethrj** [comment]
@amandalund The changes look great. Because of an infrastructure merge, your branch now conflicts. Could you rebase it or would you like me to? Also, on the updated master, I've changed the "thread ID" returned by the kernel parameter calculator to be an OpaqueId object so it can be unambiguously used. Could you change the RNG engine constructor to take a `ThreadId` instead of an `int` in its constructor a la the VGGeometry and Particle classes on the `geometry` branch? Thanks much!

**Pull request merged**